### PR TITLE
Proper retrieval of map values from storage, with bytes key

### LINKED ISSUE #631: Allow data part of log opcode to be any length of bytes
### What's your issue about?

image

If `log4(p, s, t1, t2, t3, t4)` defines the  log opcode, memory from p to p + s represents the data section of the log[0-4] instruction. t1-t4 are calls for topics (or indexed).  Since the opcode support that to be any length. Currently the un-indexed "data" parameters ( p & s parameters) of a logging operation is limited to 32 bytes in vyper.

The following should work after this fix:

```python
MyLog: __log__({arg1: indexed(bytes <= 3), arg2: bytes <= 64})

@public
def foo():
    log.MyLog('bar', 'stuuuuff')
```
### How can it be fixed?

Add support for it in vyper. Also look at updating pyethereum's `abi.py`  to interpret `bytes[n]` notation on events.

### PR DESCRIPTION
**Pull request by jacqueswww**
### - What I did

Fix for #631.  Allow retrieving values from storage, with bytearrays as keys.

### - How I did it

Struggled a bit, but then managed to use the len part of the bytearray in memory to use the proper ranges to sha3 everytime.

### - How to verify it

See test.
```python
a:num[bytes<=100000]

@public
def __init__():
    self.a["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"] = 1069
```

### CODE REVIEW

**Comment by DavidKnott** [comment]
@jacqueswww I think you meant to reference issue  #637 ?

File: viper/parser/parser_utils.py [hunk 1]
```diff
@@ -402,18 +402,19 @@ def add_variable_offset(parent, key):
...
+                    ['sha3', ['add', key.args[0].args[-1], 32], ['mload', key.args[0].args[-1]]]
                 ])
             else:
-                sub = LLLnode.from_list(['sha3', ['add', key.args[0].value, 32], ceil32(key.typ.maxlen)])
+                sub = LLLnode.from_list(['sha3', ['add', key.args[0].value, 32], ['mload', key.args[0].value]])
         else:
             subtype = typ.valuetype
             sub = base_type_conversion(key, key.typ, typ.keytype)
         if location == 'storage':
+            # import ipdb; ipdb.set_trace()
```
Comment (DavidKnott) [inline viper/parser/parser_utils.py]: Stray pdb 😄

File: tests/parser/features/test_bytes_map_keys.py [hunk 2]
```diff
@@ -93,3 +93,33 @@ def set(k: bytes <= 35, v: num):
...
+def get_it3(key: bytes<=100000) -> num:
+    return self.a[key]
+    """
+
+    c = get_contract(code)
+
+    assert c.get_it2() == 1069
+    assert c.get_it2() == 1069
+    assert c.get_it3(b"a" * 33) == 1069
+    assert c.get_it3(b"test") == 0
```
Comment (DavidKnott) [inline tests/parser/features/test_bytes_map_keys.py]: Can you add a test making sure it fails when declaration has `key: bytes  <= 50` and input is `inp: bytes <= 51`?

File: viper/parser/parser_utils.py (hunk 1 above)
Comment (jacqueswww) [reply viper/parser/parser_utils.py]: ha, thanks!

File: tests/parser/features/test_bytes_map_keys.py (hunk 2 above)
Comment (jacqueswww) [reply tests/parser/features/test_bytes_map_keys.py]: This one already exists in line 85 ;)

**Comment by jacqueswww** [comment]
@DavidKnott thanks for the review, I made the necessary changes.

**Comment by jacqueswww** [comment]
@DavidKnott can finally be merged :smile:

**Pull request merged**